### PR TITLE
feat(parquet): limit buffering when writing parquet files

### PR DESCRIPTION
**Pull request by alamb**
Closes https://github.com/influxdata/influxdb_iox/issues/7783

# TLDR
I believe this PR will
1.  Reduce the memory required for compaction significantly (it should now remain much cluster to the limits configured) and will likely increase its speed
2. *Increase* the resulting parquet files (25% bigger worst case). 
3. Note we could improve the parquet sizes by buffering more (e.g 1GB instead of 512MB per parquet file)
4. We have a plan (https://github.com/influxdata/influxdb_iox/issues/7860) that requires upstream arrow-rs changes, but can improve the parquet file size without requiring more memory

# Rationale:
As explained on https://github.com/influxdata/influxdb_iox/issues/7783#issuecomment-1557789373 the parquet writer will currently buffer up to 1M rows, prior to encoding a row group, regardless of the amount of memory that consumes. For some types of data, this amount of memory can be 10s of GB which causes the compactor to be OOM killed.

# Changes
1. Add explicit buffering (with a limit) of `RecordBatch`es to the parquet writer in IOx 
2. Add tests for same

# Effect on Memory Usage

While writing out the parquet files from the scenario on https://github.com/influxdata/influxdb_iox/issues/7783 with this change, the memory profile is consistent and stays under 1GB. Here is a screen shot from `Instruments`:
Screenshot 2023-05-23 at 5 55 49 PM

On `main`, the memory usage peaks at 20+GB (picture from https://github.com/influxdata/influxdb_iox/issues/7783#issuecomment-1557713255):
Screenshot 2023-05-22 at 2 35 11 PM

# Effect on Parquet File Size

For each parquet write, this PR buffers up to 512 MB of `RecordBatch` data, and if that is exceeded forces the writer to write out whatever it has so far. For some datasets this results in fewer rows per group and thus lower compression for the resulting parquet files.  After https://github.com/apache/arrow-rs/issues/3871 is available, the buffering can be done much more efficiently at the parquet level and we can restore the parquet file improvement

| Size of first file | Time to create | Description |
|--------|--------|--------|
| 113,430,799  | 151.25s | `main` |
| 143,247,332 (25% larger)| 71.35s | this PR | 

<details><summary>Details</summary>
<p>
Here are the sizes of the first parquet file written in the scenario  on main
```
2023-05-23T23:50:19.892374Z DEBUG parquet_file::serialize: Created parquet file num_batches=3419 num_rows=14721462 object_store_id=c4778201-26de-4980-94ee-4705a5d2f87a partition_id=PartitionId(1) write_batch_size=1024 max_row_group_size=1048576
2023-05-23T23:50:19.896574Z DEBUG parquet_file::storage: Uploading parquet to object store file_size=113430799 object_store_id=c4778201-26de-4980-94ee-4705a5d2f87a partition_id=PartitionId(1) total_time_to_create_parquet_bytes=151.246987854s
```

Here is the first parquet file made with this PR: 
```
2023-05-24T10:22:29.745603Z DEBUG parquet_file::serialize: Created parquet file num_batches=3419 num_rows=14721462 object_store_id=f5d7b027-da96-4e49-8bf7-f93673b9d1f0 partition_id=PartitionId(1) write_batch_size=1024 max_row_group_size=1048576
2023-05-24T10:22:29.810283Z DEBUG parquet_file::storage: Uploading parquet to object store file_size=143247332 object_store_id=f5d7b027-da96-4e49-8bf7-f93673b9d1f0 partition_id=PartitionId(1) total_time_to_create_parquet_bytes=71.356486275s

```

</p>
</details> 

Here is a summary on the here is the compression data. :

| Delta | Total Size | Description |
|--------|--------|--------|
| 235M | 0 | Original (non compacted) |
| 542M | 307M | `main` (without this PR, use 20GB of memory) |
| 583M | 348M  | This PR (uses les than 1GB of memory) |

This results in row groups between 16,000 and 60,000 rows per group with the reproducer data from https://github.com/influxdata/influxdb_iox/issues/7783

<details><summary>Here are some example row group sizes</summary>
<p>

Example logs:
```
2023-05-23T18:30:01.232107Z DEBUG parquet_file::writer: Attemping to reserve additional space increment=4855104 rows=13616 max_buffer_rows=1048576 buffer_size=522596128 max_buffer_size=536870912
2023-05-23T18:30:01.374408Z DEBUG parquet_file::writer: Flushing data to parquet rows=50517 max_buffer_rows=1048576 buffer_size=531001248 max_buffer_size=536870912 reservation=531001248
2023-05-23T18:30:01.440516Z DEBUG parquet_file::writer: Flushing data to parquet rows=13616 max_buffer_rows=1048576 buffer_size=522596128 max_buffer_size=536870912 reservation=522596128
2023-05-23T18:30:02.496324Z DEBUG parquet_file::writer: Flushing data to parquet rows=60708 max_buffer_rows=1048576 buffer_size=497685152 max_buffer_size=536870912 reservation=526903008
2023-05-23T18:30:03.688217Z DEBUG parquet_file::writer: Flushing data to parquet rows=61593 max_buffer_rows=1048576 buffer_size=503093696 max_buffer_size=536870912 reservation=531001248
2023-05-23T18:30:03.690626Z DEBUG parquet_file::writer: Attemping to reserve additional space increment=5732544 rows=16818 max_buffer_rows=1048576 buffer_size=528328672 max_buffer_size=536870912
2023-05-23T18:30:03.956755Z DEBUG parquet_file::writer: Flushing data to parquet rows=16818 max_buffer_rows=1048576 buffer_size=528328672 max_buffer_size=536870912 reservation=528328672
2023-05-23T18:30:05.531610Z DEBUG parquet_file::writer: Flushing data to parquet rows=60038 max_buffer_rows=1048576 buffer_size=526885376 max_buffer_size=536870912 reservation=526903008
2023-05-23T18:30:06.010282Z DEBUG parquet_file::writer: Flushing data to parquet rows=41842 max_buffer_rows=1048576 buffer_size=517718272 max_buffer_size=536870912 reservation=531001248
```

</p>
</details>

### CODE REVIEW

File: parquet_file/src/serialize.rs [hunk 1]
```diff
@@ -101,9 +120,10 @@ impl From<CodecError> for DataFusionError {
 /// [`FileMetaData`]: parquet::format::FileMetaData
 /// [`IoxParquetMetaData`]: crate::metadata::IoxParquetMetaData
 /// [`RecordBatch`]: arrow::record_batch::RecordBatch
-pub async fn to_parquet<W>(
+async fn to_parquet<W>(
     batches: SendableRecordBatchStream,
     meta: &IoxMetadata,
+    pool: Arc<dyn MemoryPool>,
```
Comment (alamb) [inline parquet_file/src/serialize.rs:126]: This is the core plumbing piece: feed the `MemoryPool` in all the way down. I tried a few different ways to hang if off the `ParquetStorage` object but that got confusing because the `Executor`, which has the `MemoryPool`  is created after the ParquetStorage in most call sites

File: parquet_file/src/serialize.rs [hunk 2]
```diff
@@ -123,15 +143,15 @@ where
 
     // Construct the arrow serializer with the metadata as part of the parquet
     // file properties.
-    let mut writer = ArrowWriter::try_new(sink, Arc::clone(&schema), Some(props))?;
+    let mut writer = TrackedMemoryArrowWriter::try_new(sink, Arc::clone(&schema), props, pool)?;
```
Comment (alamb) [inline parquet_file/src/serialize.rs:146]: This is they key change -- use a writer that encapsulate talking with the DataFusion Memory manager

File: parquet_file/src/writer.rs [hunk 3]
```diff
@@ -0,0 +1,409 @@
...
+use datafusion::{
+    error::DataFusionError,
+    execution::memory_pool::{MemoryConsumer, MemoryPool, MemoryReservation},
+};
+use observability_deps::tracing::debug;
+use parquet::{arrow::ArrowWriter, errors::ParquetError, file::properties::WriterProperties};
+use thiserror::Error;
+
+/// Default max buffer size. See [`TrackedMemoryArrowWriter::with_max_buffer_size`] for more details
+pub const MAX_BUFFER_SIZE: usize = 512 * 1024 * 1024; // 512 MB
```
Comment (alamb) [inline parquet_file/src/writer.rs:14]: This size could be increased and would result in smaller parquet files but a larger memory overhead per parquet file. We could also make it configurable, but since we have a medium term plan (https://github.com/influxdata/influxdb_iox/issues/7860) to improve buffering efficiency, I think we can avoid the extra knob for now.

**Review by Joe-Blount — APPROVED** [review]
Regarding these points:
> 2. Increase the resulting parquet files (25% bigger worst case).
> 3. Note we could improve the parquet sizes by buffering more (e.g 1GB instead of 512MB per parquet file)

If this is a concern, we could have the compactor run L2 compactions at a reduced concurrency and pass a flag to DF that means it can use more memory.  The idea is that the short lived files from early compactions remain larger & run faster, but we slow down a bit on the final pass.   We do far more L0 compactions, so the reduced concurrency would apply to a small percent of the total jobs.  But if the medium term improvement works out and settings creating larger files  doesn't stick around very long, its not worth messing with.

File: parquet_file/src/writer.rs (hunk 3 above)
Comment (alamb) [reply parquet_file/src/writer.rs:14]: I will add a comment to this effect -- I think given I have worked the buffering into the DataFusion memory pool, if we upped this constant the effect could be that the compactor might see more "out of memory" runtime errors if the memory pool size was insufficient

File: parquet_file/src/serialize.rs [hunk 4]
```diff
@@ -59,6 +64,10 @@ pub enum CodecError {
     /// Attempting to clone a handle to the provided write sink failed.
     #[error("failed to obtain writer handle clone: {0}")]
     CloneSink(std::io::Error),
+
+    /// Could not allocate sufficient memory
+    #[error("failed to allocate buffer while writing parquet: {0}")]
+    OutOfMemory(DataFusionError),
```
Comment (domodwyer) [inline parquet_file/src/serialize.rs:70]: Shouldn't this make use of the:

```rust
DataFusion(#[from] DataFusionError),
```

variant above? There's two variants that map the datafusion error, so might be a bit confusing / easy to accidentally return the wrong one via the `From` impl with `?`.

The caller can easily match on the inner error being an `OutOfMemory` error :+1:

File: parquet_file/src/serialize.rs [hunk 5]
```diff
@@ -154,6 +174,7 @@ where
 pub async fn to_parquet_bytes(
     batches: SendableRecordBatchStream,
     meta: &IoxMetadata,
+    pool: Arc<dyn MemoryPool>,
```
Comment (domodwyer) [inline parquet_file/src/serialize.rs:177]: Minor, but it's a shame we can't pass in an option/none for cases where we don't care - a fair few bits of this diff don't! I can't think of a way to express it nicely though because of the writer types being different.

File: parquet_file/src/writer.rs [hunk 6]
```diff
@@ -0,0 +1,409 @@
...
+            rows = self.buffer_rows,
+            max_buffer_rows = self.max_buffer_rows,
+            buffer_size = self.buffer_size,
+            max_buffer_size = self.max_buffer_size,
+            reservation = self.reservation.size(),
+            "Flushing data to parquet"
+        );
+
+        while let Some(batch) = self.buffer.pop_front() {
+            self.inner.write(&batch)?;
```
Comment (domodwyer) [inline parquet_file/src/writer.rs:223]: If the write fails, and the flush partially completes, the tracking vars (`buffer_rows` and `buffer_size` aren't updated - perhaps they need lifting into the loop?

We could then `assert_eq!(self.buffer_rows, 0)'` at the end of flushing too, to assert correct tracking :+1:

File: parquet_file/src/writer.rs [hunk 7]
```diff
@@ -0,0 +1,409 @@
...
+
+        // enforce the limits, if needed
+        if self.buffer_size + batch_size > self.max_buffer_size
+            || self.buffer_rows + batch_rows > self.max_buffer_rows
+        {
+            self.flush()?;
+        }
+        self.buffer_rows += batch_rows;
+        self.buffer_size += batch_size;
+        self.ensure_reservation()?;
```
Comment (domodwyer) [inline parquet_file/src/writer.rs:178]: Likewise here, if the reservation fails, the tracking vars are left out of whack w.r.t the contents of the buffer.

**Review by domodwyer — DISMISSED** [review]
It's a shame we can't embed this into the `ParquetStorage` type rather than passing it in each time - it'd be much cleaner that way!

I think the tracking vars are not correctly maintained in the presence of errors.

Would you expect any change to the ingester memory usage with this?

File: parquet_file/src/serialize.rs (hunk 4 above)
Comment (alamb) [reply parquet_file/src/serialize.rs:70]: Sure, I can make that change. I figured an explicit error type would make it clear to a caller that the only DataFusion error that is expected is an out of memory one. Instead I will add it as a comment

File: parquet_file/src/serialize.rs (hunk 5 above)
Comment (alamb) [reply parquet_file/src/serialize.rs:177]: 👍  I also think leaving it explicit forces the caller to know there is something down the stack that requires allocating memory.

File: parquet_file/src/writer.rs (hunk 7 above)
Comment (alamb) [reply parquet_file/src/writer.rs:178]: https://github.com/influxdata/influxdb_iox/pull/7855#discussion_r1204496318

**Comment by alamb** [comment]
> Would you expect any change to the ingester memory usage with this?

@domodwyer  No, there should be no actual memory usage change (as the arrow-rs parquet writer is already doing this buffering: https://docs.rs/parquet/40.0.0/src/parquet/arrow/arrow_writer/mod.rs.html#144-146 )

However, what could happen is that if the ingester tries to persist a parquet file (or a lot of parquet files concurrently) that would consume more than its configured memory pool size, it may get a runtime error.

File: parquet_file/src/writer.rs (hunk 7 above)
Comment (alamb) [reply parquet_file/src/writer.rs:178]: It is a  good call -- I will fix this and add a test

File: parquet_file/src/writer.rs [hunk 8]
```diff
@@ -0,0 +1,409 @@
...
+        // add batch to queue
+        self.buffer.push_back(batch);
+        Ok(())
+    }
+
+    /// Ensures that our reservation with the memory pool is
+    /// sufficiently large for the amount of data that is currently
+    /// buffered. If not, attempts to grow the reservation to match
+    /// what is currently buffered.
+    pub fn ensure_reservation(&mut self) -> Result<()> {
```
[the remark this answers is not in the document]
Comment (alamb) [reply parquet_file/src/writer.rs]: > I'm not sure I follow the need for min_allocation_increment but otherwise it looks correct to me

I was trying to avoid many small allocation calls. Maybe it doesn't really matter 🤔

**Comment by alamb** [comment]
> > Regarding these points:

> Increase the resulting parquet files (25% bigger worst case).
> Note we could improve the parquet sizes by buffering more (e.g 1GB instead of 512MB per parquet file)

> If this is a concern, we could have the compactor run L2 compactions at a reduced concurrency and pass a flag to DF that means it can use more memory. The idea is that the short lived files from early compactions remain larger & run faster, but we slow down a bit on the final pass. We do far more L0 compactions, so the reduced concurrency would apply to a small percent of the total jobs. But if the medium term improvement works out and settings creating larger files doesn't stick around very long, its not worth messing with.

I agree with this assessment @Joe-Blount. 👍

File: parquet_file/src/serialize.rs (hunk 4 above)
Comment (alamb) [reply parquet_file/src/serialize.rs]: b59702de9

File: parquet_file/src/writer.rs (hunk 3 above)
Comment (alamb) [reply parquet_file/src/writer.rs:20]: 2d51fb238

File: parquet_file/src/writer.rs (hunk 6 above)
Comment (alamb) [reply parquet_file/src/writer.rs:234]: 9a479a8b5

**Comment by alamb** [comment]
> Can we pre-stage the relevant config in prod so that we are using a sufficiently large limit to avoid this? The ingester is (understandably) not designed to cope with serialisation failing in this way.

Good call. I did some research and I don't think we need to make any additional changes. 

The affected setting is `INFLUXDB_IOX_EXEC_MEM_POOL_BYTES`, which defaults to `USIZE::MAX` (unlimited) for the `ingester`:

https://github.com/influxdata/influxdb_iox/blob/89d8207784de96646f92aac7603e199754b3d0e9/influxdb_iox/src/commands/run/ingester.rs#L78-L85C1

```shell
influxdb_iox run ingester --help
...
      --exec-mem-pool-bytes <EXEC_MEM_POOL_BYTES>
          Size of memory pool used during query exec, in bytes

          [env: INFLUXDB_IOX_EXEC_MEM_POOL_BYTES=]
          [default: 18446744073709551615]
```

The ingester configuration on production does not appear to override this setting. https://github.com/search?q=repo%3Ainfluxdata%2Fk8s-idpe%20INFLUXDB_IOX_EXEC_MEM_POOL_BYTES&type=code

For example, 
https://github.com/influxdata/k8s-idpe/blob/de4b56f0960f7bb588369d8047c4a2f139d6abe9/env/aws/cst-prod01-us-east-1/iox/yml/apps-v1.StatefulSet-iox.iox-shared-ingester.yml

**Comment by domodwyer** [comment]
Awesome - thanks for checking - I believe our current config doesn't override this anywhere :+1:

**Comment by alamb** [comment]
Update here is that we have a potential alternate solution in https://github.com/influxdata/influxdb_iox/pull/7866 -- see [internal slack link](https://influxdata.slack.com/archives/CRK9M8L5Q/p1685022677628069)

We will spend some time evaluating that and then decide if we should proceed with merging this PR or if we should do something else

**Comment by alamb** [comment]
Please see updated on https://github.com/influxdata/influxdb_iox/issues/7783#issuecomment-1563434629
